### PR TITLE
Binding operators in occurrence and locate

### PR DESCRIPTION
**Pull request by mattiase**
Find binding operator uses in Occurrence and their definitions in Locate (issue #1398)
Although Locate works from a position where the binding operator syntax is used, Occurrence does not but requires the position to indicate the reified (bracketed) operator.

### CODE REVIEW

File: src/ocaml/preprocess/lexer_ident.mll [hunk 1]
```diff
@@ -77,6 +77,8 @@ rule token = parse
       { OPTLABEL label }
   | "?" (lowercase_latin1 identchar_latin1 *) as label ':'
       { OPTLABEL label }
+  | ("let" (symbolchar +)) as op { LETOP op }
+  | ("and" (symbolchar +)) as op { ANDOP op }
```
Comment (voodoos) [inline src/ocaml/preprocess/lexer_ident.mll:81]: This looks slightly incorrect. In the compiler's lexer, the rule is:
```
| "let" kwdopchar dotsymbolchar * as op
```
with
```
let kwdopchar = ['$' '&' '*' '+' '-' '/' '<' '=' '>' '@' '^' '|']
let dotsymbolchar = ['!' '$' '%' '&' '*' '+' '-' '/' ':' '=' '>' '?' '@' '^' '|']
```

But here `symbolchar` is bigger than `kwdopchar`:
```
['!' '$' '%' '&' '*' '+' '-' '.' '/' ':' '<' '=' '>' '?' '@' '^' '|' '~']
```

So this rule matches more operators than allowed.

File: src/ocaml/merlin_specific/browse_raw.ml [hunk 2]
```diff
@@ -769,10 +769,16 @@ let module_expr_paths { Typedtree. mod_desc } =
     [reloc (Path.Pident id) loc, Option.map mk_lident loc.txt]
   | _ -> []
 
+let bindop_path { bop_op_name; bop_op_path } =
+  let loc = bop_op_name
+  and path = bop_op_path in
```
Comment (voodoos) [inline src/ocaml/merlin_specific/browse_raw.ml:774]: Nitpick: we generally do not use the `and` keyword for non-mutually-recursive definitions in Merlin's codebase. Multiple `let ... in` are preferred.

File: tests/test-dirs/occurrences/issue1398.t/run.t [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        "end": {
+          "line": 4,
+          "col": 17
+        }
+      }
+    ],
+    "notifications": []
+  }
+
+Test finding occurrences of let-based binding operator, from operator syntax:
```
Comment (voodoos) [inline tests/test-dirs/occurrences/issue1398.t/run.t:81]: Can you add a FIXME annotation before tests that have an unexpected result.

**Review by voodoos — COMMENTED** [review]
Thank you for that contribution @mattiase 

Apart from a few small things it looks good to me. 
There is still some issues with occurrences but it already improves the situation a lot.

Can you rebase and add a Changelog entry ?

File: src/ocaml/preprocess/lexer_ident.mll (hunk 1 above)
Comment (mattiase) [reply src/ocaml/preprocess/lexer_ident.mll]: Right, it was an over-approximation. Now it should be precise.

File: src/ocaml/merlin_specific/browse_raw.ml (hunk 2 above)
Comment (mattiase) [reply src/ocaml/merlin_specific/browse_raw.ml]: Fixed.

File: tests/test-dirs/occurrences/issue1398.t/run.t (hunk 3 above)
Comment (mattiase) [reply tests/test-dirs/occurrences/issue1398.t/run.t:82]: Done.

File: CHANGES.md [hunk 4]
```diff
@@ -10,6 +10,7 @@ git version
     - handle `=` syntax in compiler flags (#1409)
     - expose all destruct exceptions in the api (#1437)
     - fix superfluous break in error reporting (#1432)
+    - recognise binding operators in locate and occurrences (#1398)
```
Comment (voodoos) [inline CHANGES.md:13]:
```suggestion
    - recognise binding operators in locate and occurrences (#1398, @mattiase)
```

File: src/ocaml/merlin_specific/browse_raw.ml [hunk 5]
```diff
@@ -776,10 +776,17 @@ let module_expr_paths { Typedtree. mod_desc } =
...
+  let loc = bop_op_name in
+  let path = bop_op_path in
+  (reloc path loc, Some (Longident.Lident loc.txt))
+
 let expression_paths { Typedtree. exp_desc; exp_extra; exp_env; _ } =
   let init =
     match exp_desc with
     | Texp_ident (path,loc,_) -> [reloc path loc, Some loc.txt]
+    | Texp_letop {let_; ands} ->
+       bindop_path let_ :: List.map ~f:bindop_path ands
```
Comment (voodoos) [inline src/ocaml/merlin_specific/browse_raw.ml:789]:
```suggestion
    | Texp_letop {let_; ands} ->
      bindop_path let_ :: List.map ~f:bindop_path ands
```

(match indentation used in the rest of the function)

Comment (mattiase) [reply src/ocaml/merlin_specific/browse_raw.ml:789]: Good catch, thank you – looks like Tuareg has a different idea of the indentation here but clearly the existing usage is to be followed.

File: CHANGES.md (hunk 4 above)
Comment (mattiase) [reply CHANGES.md:13]: Not really fond of tooting my own horn but you make the rules; changing.

Comment (voodoos) [reply CHANGES.md]: We usually credit external contributors, but don't feel obliged to do it if you prefer to remain in the shadows :-)

Comment (voodoos) [reply CHANGES.md]: > We usually credit external contributors, but don't feel obliged to do it if you prefer to remain in the shadows :-)

**Comment by voodoos** [comment]
Not sure why the CI broke (but it is most probably not related to this PR).
@smorimoto could that be related to the release of 4.14 ? But our CI config didn't change and ask for 4.13: 
```
ocaml-version:
          - 4.13.0
          ```

The error is:
```
Install system packages required by opam packages
  /Users/runner/hostedtoolcache/opam/2.1.2/x86_64/opam depext dot-merlin-reader merlin
  # Detecting depexts using vars: arch=x86_64, os=macos, os-distribution=homebrew, os-family=homebrew
  [ERROR] No solution for dot-merlin-reader & merlin:   * No agreement on the version of ocaml-base-compiler:
              - (invariant) → ocaml-base-compiler >= 4.14.0
              - merlin → ocaml < 4.14 → ocaml-base-compiler = 3.07+2
              You can temporarily relax the switch invariant with `--update-invariant'
  
  
  'opam --cli=2.1 list --readonly --external --resolve=dot-merlin-reader,merlin' failed.
  Command failed: opam --cli=2.1 list --readonly --external '--resolve=dot-merlin-reader,merlin' returned 20
  Error: The process '/Users/runner/hostedtoolcache/opam/2.1.2/x86_64/opam' failed with exit code 20
```

Apparently OCaml 4.14 is installed even if our workflow file specify 4.13...
```
Install OCaml
  /Users/runner/hostedtoolcache/opam/2.1.2/x86_64/opam switch create . --no-install --packages ocaml-base-compiler.4.14.0
  ```

**Comment by smorimoto** [comment]
I will raise a PR to fix that issue.

**Pull request merged**